src/SprykerEco/Zed/Payone/Business/Distributor/PriceDistributor.php

```diff
@@ -0,0 +1,122 @@
+    {
+        if ($orderTransfer->getPayments()->count() <= 1) {
+            return $orderTransfer;
+        }
+
+        $payonePayment = $this->findPayonePayment($orderTransfer);
+        if ($payonePayment === null) {
+            return $orderTransfer;
+        }
+
+        $totals = $orderTransfer->getTotals();
+
+        $roundingError = 0;
+        $priceRatio = $payonePayment->getAmount() / $totals->getGrandTotal();
+
+        $this->distributeOrderItemsPrices($orderTransfer, $priceRatio, $roundingError);
+        $this->distributeOrderExpensesPrices($orderTransfer, $priceRatio, $roundingError);
+
+        $orderTransfer->setTotals($totals->setGrandTotal($payonePayment->getAmount()));
+
+        return $orderTransfer;
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     *
+     * @return \Generated\Shared\Transfer\PaymentTransfer|null
+     */
+    protected function findPayonePayment(OrderTransfer $orderTransfer): ?PaymentTransfer
+    {
+        $orderTransfer->getPayments();
+        foreach ($orderTransfer->getPayments() as $payment) {
+            if ($payment->getPaymentProvider() === static::PAYMENT_PROVIDER_NAME) {
+                return $payment;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param int $unitPrice
+     * @param float $priceRatio
+     * @param float $roundingError
+     *
+     * @return int
+     */
+    protected function calculateRoundedPrice(int $unitPrice, float $priceRatio, float &$roundingError): int
+    {
+        $priceBeforeRound = ($unitPrice * $priceRatio) + $roundingError;
+        $priceRounded = (int)round($priceBeforeRound);
+        $roundingError = $priceBeforeRound - $priceRounded;
+
+        return $priceRounded;
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     * @param float $priceRatio
+     * @param float $roundingError
+     *
+     * @return void
+     */
+    protected function distributeOrderItemsPrices(OrderTransfer $orderTransfer, float $priceRatio, float &$roundingError): void
+    {
+        foreach ($orderTransfer->getItems() as $itemTransfer) {
+            $priceRounded = $this->calculateRoundedPrice(
+                $itemTransfer->getSumPriceToPayAggregation(),
+                $priceRatio,
+                $roundingError,
+            );
+
+            $itemTransfer->setSumPriceToPayAggregation($priceRounded);
+
+            $unitPrice = $priceRounded / $itemTransfer->getQuantity();
+            $itemTransfer->setUnitPriceToPayAggregation($unitPrice);
+            $itemTransfer->setUnitGrossPrice($unitPrice);
+        }
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     * @param float $priceRatio
+     * @param float $roundingError
+     *
+     * @return void
+     */
+    protected function distributeOrderExpensesPrices(OrderTransfer $orderTransfer, float $priceRatio, float &$roundingError): void
```

Comment:
Extract the method to the class `OrderExpensePriceDistributor`.

Reply:
Discussed